packages/components/src/components/TokenGate/TokenGate.tsx

```diff
@@ -0,0 +1,83 @@
+import React, { useEffect, useState, ReactNode } from 'react';
+import { ethers } from 'ethers';
+import { Spinner } from '@chakra-ui/react';
+import { ERC20ABI } from '@web3-ui/hooks/src/constants';
+export interface TokenGateProps {
+  /**
+   * The provider or signer to fetch the address from the ens
+   */
+  provider: ethers.providers.Web3Provider;
+  /**
+   * The token contract address
+   */
+  contractAddress: string;
+  /**
+   * The token quantity required to access child component. Default=1
+   */
+  requiredQuantity?: number;
+  /**
+   * Child nodes
+   */
+  children: ReactNode;
+  /**
+   * Optional message if access denied
+   */
+  deniedMessage?: ReactNode;
+  /**
+   * Optional spinner during loading
+   */
+  loader?: boolean;
+}
+export const TokenGate: React.FC<TokenGateProps> = ({
+  contractAddress,
+  provider,
+  requiredQuantity = 1,
+  children,
+  deniedMessage,
+  loader = false,
+}) => {
+  const [tokenQuantity, setTokenQuantity] = useState<number>(0);
+  const [loadedStatus, setloadedStatus] = useState<boolean>(false);
+  // connect to contract address to get balance
+  async function getTokenBalance() {
+    const signer = provider!.getSigner();
```

Comment:
1. Do we need a signer to read data?
2. If we _do_ need a signer in all cases, what do you think of accepting a `signer` as a prop instead of a `provider`?

Reply:
That would work. But should we take the signer at all? Or should we allow the dev to pass the quantity held by the wallet? 

In the core package it would use then use the get balance hook to provide that for the dev. 

Maybe the dev has the wallet balance in their app's state already so fetching the balance for them could be undesirable.

And if they want the balance fetched for them they can use the core package version.

Reply:
> In the core package it would use the get balance hook to provide that for the dev. Maybe the dev has the wallet balance in their app's state already so fetching the balance for them could be undesirable.

With you on this. Let's just take the balance as a prop :)

Reply:
@hone1er do you also want to add Token Gate to `core` in this PR?